libyara/modules/macho.c

```diff
@@ -0,0 +1,1225 @@
+  else                                                                         \
+  {                                                                            \
+    set_integer(offset, object, "entry_point");                                \
+  }                                                                            \
+  set_integer(yr_##bo##64toh(ep_command->stacksize), object, "stack_size");    \
+}                                                                              \
+
+MACHO_HANDLE_MAIN(le)
+MACHO_HANDLE_MAIN(be)
+
+
+// Load segment and its sections.
+
+#define MACHO_HANDLE_SEGMENTS_SECTIONS(bits,bo)                                \
+void macho_handle_segments_sections_##bits##_##bo(                             \
+    void* command,                                                             \
+    uint32_t i,                                                                \
+    YR_OBJECT* object)                                                         \
+{                                                                              \
+  segment_command_##bits##_t* sg = (segment_command_##bits##_t*)command;       \
+                                                                               \
+  set_sized_string(sg->segname, strnlen(sg->segname, 16),                      \
+                   object, "segments[%i].segname", i);                         \
+                                                                               \
+  set_integer(yr_##bo##bits##toh(sg->vmaddr),                                  \
+              object, "segments[%i].vmaddr", i);                               \
+  set_integer(yr_##bo##bits##toh(sg->vmsize),                                  \
+              object, "segments[%i].vmsize", i);                               \
+  set_integer(yr_##bo##bits##toh(sg->fileoff),                                 \
+              object, "segments[%i].fileoff", i);                              \
+  set_integer(yr_##bo##bits##toh(sg->filesize),                                \
+              object, "segments[%i].fsize", i);                                \
+  set_integer(yr_##bo##32toh(sg->maxprot),                                     \
+              object, "segments[%i].maxprot", i);                              \
+  set_integer(yr_##bo##32toh(sg->initprot),                                    \
+              object, "segments[%i].initprot", i);                             \
+  set_integer(yr_##bo##32toh(sg->nsects),                                      \
+              object, "segments[%i].nsects", i);                               \
+  set_integer(yr_##bo##32toh(sg->flags),                                       \
+              object, "segments[%i].flags", i);                                \
+                                                                               \
+  for (unsigned j = 0; j < yr_##bo##32toh(sg->nsects); ++j)                    \
```

Comment:
No sanity checks for preventing out-of-bounds reads. If `sg->nsects` is altered or the file is truncated after the header this will segfault.